### ISSUE TITLE
docker based action not pulling latest base image when building container

### ISSUE THREAD

**Issue by thexperiments** [actions/runner#1713]
**Describe the bug**
When using a docker based action the runner builds the container for that action based on the Dockerfile but it doesn't check if there is a new image with the tag specified in FROM in the registry.

**To Reproduce**
1. have a Dockerfile based action and run it
2. change the base image referenced in the Dockerfile
3. run the action again

**Expected behavior**
action is executed in a container based on the changed base container

## Runner Version and Platform
Runner: 2.278.0
GitHub Enterprise Server 3.1.14

Runner OS Ubuntu 20.04

## What's not working?
old, cached base image used

**Comment by ruvceskistefan** [comment]
Hi @thexperiments,
Thanks for opening this issue! We will try to reproduce it to understand the cause of the issue and we will get back to you as soon as possible.

**Comment by nikola-jokic** [comment]
Hi, @thexperiments,

While trying to reproduce this issue, I did find that it works fine. Could you please explain how did you figure out that the previous image is being used, instead of the new one which has a different `FROM` instruction?

**Comment by thexperiments** [comment]
Hi, what I did was to trigger the action again after I did a change to the container used in the from statement.
As I had added a tool that is used by the action the action failed.

I logged into the runner via ssh and did a docker pull on the image which resulted in it being updated. 
After I did this the action succeeded without a problem on that runner triggered in the same way as before.

This tells me: 
* from registry everything was fine and the new version was published properly
* the image had not been pulled/updated on that worker previously

**Comment by nikola-jokic** [comment]
The way it works is that, when action is triggered, the runner will execute the `docker build` command with the tag set. That tag is later re-used for the next builds of the same directory, but if you have changed the `FROM` instruction, from a workflow, you can see that the new docker image has been built, but the tag will stay the same. If the `docker build` command succeeds on the first attempt and fails on the second one, the tag would still be in the build cache.

**Comment by thexperiments** [comment]
The thing is that the *FROM* section is not changed, it is pointing to the *latest* tag. Only the image that the *latest* tag is referring to is updated in the registry.

Thus the container build could use `docker build --pull` so that the build process for the container always checks if there is a new image for the used tag.

I would not use the *latest* tag in a production environment but the same would be true for referencing a major version tag which regularly get "reassigned" to updated images.

**Comment by nikola-jokic** [comment]
Oh, I get it now. I will discuss this with the team and I will get back to you 😊. Thanks for clarifying!

**Comment by nikola-jokic** [comment]
Hey @thexperiments,

Sure, I will elaborate. This issue I have linked is describing the action on the pull with the latest tag. While we work on it, it is clear that we would have to add some logic to the `:latest` tag so that is why I have closed it. But after some thought, I will re-open it and mark this as an enhancement as well and I will add it to the board, but the same fix would apply to both issues.

While discussing with the team, it seems like it would be preferred not to use `Dockerfile` based action, but we will cover this use-case in the future as well. The preferred way would be to use 

Thank you for your quick response!

**Comment by thexperiments** [comment]
Thanks for the quick response and reopening.
Dockerfile is an important part of making actions work for us. We need to be able to also runn all CI parts without GitHUb which a Dockerfile that we can also run and build outside of GitHub provides us the necessary abstraction. In addition to this the people maintaining the action and the people maintaining the workflow files are not the same.

Thus we don't want anything specified in the workflows on what container to use etc.

**Comment by nikola-jokic** [comment]
Of course! Sorry for closing it prematurely. I will keep you posted on this one.

**Comment by fno2010** [comment]
Any follow-up on this issue? At least, I hope there is a button so that we can clean up the runner storage and re-run the action with a clean slate. Just similar to the clear cache feature in travis-ci.

I don't know where a docker image is exactly stored for a repo. I assume it is repo specific? Because when I create a new repo, the action will pull the latest docker image. I hope there can be some way to treat an existing repo as a new one when running an action.

**Comment by JohnKiller** [comment]
This is still needed. Please remove the stale label.
